pkg/modifycontroller/controller.go

```diff
@@ -0,0 +1,378 @@
+			pvcKey, err := cache.MetaNamespaceKeyFunc(pvc)
+			if err != nil {
+				return err
+			}
+			ctrl.uncertainPVCs[pvcKey] = *pvc.DeepCopy()
+		}
+	}
+
+	return nil
+}
+
+func (ctrl *modifyController) addPVC(obj interface{}) {
+	objKey, err := getObjectKey(obj)
+	if err != nil {
+		return
+	}
+	ctrl.claimQueue.Add(objKey)
+}
+
+func (ctrl *modifyController) updatePVC(oldObj, newObj interface{}) {
+	oldPVC, ok := oldObj.(*v1.PersistentVolumeClaim)
+	if !ok || oldPVC == nil {
+		return
+	}
+
+	newPVC, ok := newObj.(*v1.PersistentVolumeClaim)
+	if !ok || newPVC == nil {
+		return
+	}
+
+	// Only trigger modify volume if the following conditions are met
+	// 1. Feature gate is enabled
+	// 2. Non empty vac name
+	// 3. oldVacName != newVacName
+	// 4. PVC is in Bound state
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeAttributesClass) {
+		oldVacName := oldPVC.Spec.VolumeAttributesClassName
+		newVacName := newPVC.Spec.VolumeAttributesClassName
+		if newVacName != nil && *newVacName != "" && *newVacName != *oldVacName && oldPVC.Status.Phase == v1.ClaimBound {
+			volumeObj, exists, err := ctrl.volumes.GetByKey(oldPVC.Spec.VolumeName)
+			if err != nil {
+				klog.Errorf("Get PV %q of pvc %q failed: %v", oldPVC.Spec.VolumeName, klog.KObj(oldPVC), err)
+				return
+			}
+			if !exists {
+				klog.InfoS("PV bound to PVC not found", "PV", oldPVC.Spec.VolumeName, "PVC", klog.KObj(oldPVC))
+				return
+			}
+
+			pv, ok := volumeObj.(*v1.PersistentVolume)
+			if !ok {
+				klog.Errorf("expected volume but got %+v", volumeObj)
+				return
+			}
+
+			_, _, err, _ = ctrl.modify(newPVC, pv)
+			if err != nil {
+				klog.Errorf("failed to modify volume at pvc update %v", err)
+				return
+			}
+		} else {
+			klog.V(4).InfoS("No need to modify PVC", "PVC", klog.KObj(newPVC))
+		}
+	}
+}
+
+func (ctrl *modifyController) deletePVC(obj interface{}) {
+	objKey, err := getObjectKey(obj)
+	if err != nil {
+		return
+	}
+	ctrl.claimQueue.Forget(objKey)
+}
+
+// Patches a given PVC with changes from newPVC. If addResourceVersionCheck is true
+// then a version check is added to the patch to ensure that we are not patching
+// old(and possibly outdated) PVC objects.
+func (ctrl *modifyController) patchClaim(oldPVC, newPVC *v1.PersistentVolumeClaim, addResourceVersionCheck bool) (*v1.PersistentVolumeClaim, error) {
+	patchBytes, err := util.GetPVCPatchData(oldPVC, newPVC, addResourceVersionCheck)
+	if err != nil {
+		return oldPVC, fmt.Errorf("can't patch status of PVC %s as generate path data failed: %v", klog.KObj(oldPVC), err)
+	}
+	updatedClaim, updateErr := ctrl.kubeClient.CoreV1().PersistentVolumeClaims(oldPVC.Namespace).
+		Patch(context.TODO(), oldPVC.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{}, "status")
+	if updateErr != nil {
+		return oldPVC, fmt.Errorf("can't patch status of  PVC %s with %v", klog.KObj(oldPVC), updateErr)
+	}
+	err = ctrl.claims.Update(updatedClaim)
```

Comment:
why we need change informer store? Isn't it dangerous?

Reply:
This is the same as the resizer, if we update the info of resizing operation, we should update modify volume too.

@gnufied is it dangerous?

Reply:
Controllers like PV-controller keep an additional cache in front of the informer cache. @jsafrane do you remember the reason?

Reply:
@gnufied Do you know the reason?

Reply:
its been awhile, but IIRC the reason was - it would allow controller to read updated object immediately and not cause unnecessary updates. But its been awhile....

Reply:
Having said that, I don't think it is dangerous.

>Also especially because this is a shared informer, that the informer events won't get triggered for the other controller.

Are you saying - if resize controller updates the pvc and updates it in store, then modify controller will not receive an event?

Reply:
Yes discussing with some api-machinery folks, they highly recommend not updating the informer store because it could cause inconsistency issues like this.

The best practice is to make the controller design eventually consistent and be able to handle potentially stale objects.